cmd/ebpf/xdp/sdf_filter.h

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright 2023 Edgecom LLC
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+#include <linux/bpf.h>
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+
+#include "xdp/utils/trace.h"
+#include "xdp/utils/packet_context.h"
+#include "xdp/utils/types.h"
+#include "xdp/pdr.h"
+
+struct ip_subnet {
+    __u8 type; // 0: any, 1: ip4, 2: ip6
+    // If type != any, ip field has meaningful value.
+    // If IPv4 -> lower 32 bits. If IPv6 -> all 128 bits.
+    __u128 ip;
+    // If type != any, mask field has meaningful value.
+    // If IPv4 mask -> lower 32 bits. If IPv6 mask -> all 128 bits.
+    // Should always be applied to matching ip (except type == any).
+    __u128 mask;
+};
+
+struct port_range {
+    __u16 lower_bound; // If not specified in SDF: 0
+    __u16 upper_bound; // If not specified in SDF: 65535
+};
+
+struct sdf_filter {
+    __u8 protocol; // Required by SDF. 0: icmp, 1: ip, 2: tcp, 3: udp
+    struct ip_subnet src_addr;
+    struct port_range src_port;
+    struct ip_subnet dst_addr;
+    struct port_range dst_port;
+};
+
+
+static __always_inline __u8 get_sdf_protocol(__u8 ip_protocol) {
+    switch(ip_protocol)
+    {
+        case IPPROTO_ICMP: return 0;
+        case IPPROTO_TCP: return 2;
+        case IPPROTO_UDP: return 3;
+        default: return 1;
+    }
+}
+
+static __always_inline __u8 match_sdf_filter_ipv4(const struct packet_context *ctx, const struct sdf_filter *sdf) {
+    if(!ctx || !ctx->ip4 || !sdf)
+        return 0;
+
+    const struct iphdr *ip4 = ctx->ip4;
+    __u8 packet_protocol = get_sdf_protocol(ip4->protocol); //TODO: convert protocol in golang part
+    __u16 packet_src_port = 0;
+    __u16 packet_dst_port = 0;
+    if(ctx->udp) {
+        packet_src_port = bpf_ntohs(ctx->udp->source); //TODO: convert port in golang part
+        packet_dst_port = bpf_ntohs(ctx->udp->dest); //TODO: convert port in golang part
+    } else if (ctx->tcp) {
+        packet_src_port = bpf_ntohs(ctx->tcp->source); //TODO: convert port in golang part
+        packet_dst_port = bpf_ntohs(ctx->tcp->dest); //TODO: convert port in golang part
+    } 
```

Comment:
This comment is somewhat related to the comment on the parse_l4 function.

If a packet will come through GTP with L4 being something different from TCP and UDP, an unexpected behaviour can appear. Function match_sdf_filter_ipv4 asserts that ctx variable will have either udp or tcp field. Maybe it will be better to return 0 code if none of variables is present? I mean, if the code is written correctly, and everywhere before calling match_sdf_filter_ipv4 a function parse_l4 will be called, it is OK (because it checks available protocols). But what if it's not?

Same for function match_sdf_filter_ipv6.

Reply:
If neither tcp, nor udp is in packet, 0-value ports are used. For example, in case of icmp it causes to ignore port matching,

Reply:
But 0-value port is a possible value in packet (even if it is a system port and shouldn't be used in packets).

What about such test case:
1. SDF filter is set with flowDescription "permit out icmp from 10.62.0.1 to 8.8.8.8 0-20".
2. An icmp packet is sent from 10.62.0.1 to 8.8.8.8 with dest port = 40. 

In this case in function match_sdf_filter_ipv4 variable packet_dst_port will be =0. And 0 port is allowed by SDF filter, so packet will match SDF filter. Although initially the packet does not match the filter.

Reply:
ICMP doesn't use ports